File: stripe.py
To integrate Stripe as the payment processor for your lottery/betting application,
you need to set up Stripe's API to handle payments. Stripe is a global payment processing
platform that can handle international transactions, making it suitable for your use case.

Steps to Integrate Stripe

    Create a Stripe Account:
        Sign up for a Stripe account and obtain your API keys from the Stripe Dashboard.

    Install Required Libraries:
        Install the Stripe Python library using pip: pip install stripe.

    Configure Environment Variables:
        Store your Stripe API keys securely in environment variables.

    Implement Stripe API Integration:
        Set up functions to create payment intents and handle webhooks.

Payment Integration Implementation

config.py:


import os

class Config:
    SECRET_KEY = os.environ.get('SECRET_KEY') or 'your_secret_key'
    SQLALCHEMY_DATABASE_URI = os.environ.get('DATABASE_URL') or 'sqlite:///lottery.db'
    SQLALCHEMY_TRACK_MODIFICATIONS = False
    
    # Stripe API Configuration
    STRIPE_PUBLIC_KEY = os.environ.get('STRIPE_PUBLIC_KEY')
    STRIPE_SECRET_KEY = os.environ.get('STRIPE_SECRET_KEY')

stripe_payment.py:


import stripe
from flask import current_app, url_for

def create_payment_intent(amount, currency='usd'):
    stripe.api_key = current_app.config['STRIPE_SECRET_KEY']
    
    try:
        payment_intent = stripe.PaymentIntent.create(
            amount=amount,
            currency=currency,
            payment_method_types=['card'],
        )
        return payment_intent
    except stripe.error.StripeError as e:
        # Handle errors
        print(f"Stripe error: {str(e)}")
        return None

def handle_webhook(request):
    # Retrieve the event by verifying the signature using the raw body and secret.
    payload = request.data
    sig_header = request.headers.get('Stripe-Signature')
    endpoint_secret = current_app.config['STRIPE_ENDPOINT_SECRET']

    try:
        event = stripe.Webhook.construct_event(
            payload, sig_header, endpoint_secret
        )
    except ValueError as e:
        # Invalid payload
        print("Invalid payload")
        return None
    except stripe.error.SignatureVerificationError as e:
        # Invalid signature
        print("Invalid signature")
        return None

    # Handle the event
    if event['type'] == 'payment_intent.succeeded':
        payment_intent = event['data']['object']
        # Fulfill the purchase, mark entry as paid
        # Update your database accordingly
        print(f"PaymentIntent for {payment_intent['amount']} was successful!")

    # Other event types can be handled here

    return event



Using the Payment Processor in Your Application

routes.py:



import uuid
from flask import flash, render_template, request, redirect, url_for
from .stripe_payment import create_payment_intent, handle_webhook

@bp.route('/game/<int:game_id>/enter', methods=['GET', 'POST'])
def enter_game(game_id):
    game = Game.query.get_or_404(game_id)

    if request.method == 'POST':
        amount = request.form.get('amount')
        currency = request.form.get('currency', 'usd')
        payment_intent = create_payment_intent(int(amount) * 100, currency)

        if payment_intent:
            return render_template('checkout.html', client_secret=payment_intent.client_secret, stripe_public_key=current_app.config['STRIPE_PUBLIC_KEY'])
        else:
            flash('Failed to create payment intent.', 'danger')

    return render_template('enter_game.html', game=game)

@bp.route('/webhook', methods=['POST'])
def webhook():
    event = handle_webhook(request)
    if event:
        return '', 200
    else:
        return 'Webhook error', 400





Frontend: Handling Payment

You need to set up the frontend to handle Stripe payments using Stripe.js.

templates/checkout.html:


<!DOCTYPE html>
<html>
<head>
    <title>Checkout</title>
    <script src="https://js.stripe.com/v3/"></script>
</head>
<body>
    <h1>Complete your payment</h1>
    <form id="payment-form">
        <div id="card-element"></div>
        <button id="submit">Pay</button>
        <div id="error-message"></div>
    </form>

    <script>
        const stripe = Stripe('{{ stripe_public_key }}');
        const elements = stripe.elements();
        const cardElement = elements.create('card');
        cardElement.mount('#card-element');

        const form = document.getElementById('payment-form');

        form.addEventListener('submit', async (event) => {
            event.preventDefault();

            const { paymentIntent, error } = await stripe.confirmCardPayment(
                '{{ client_secret }}', {
                    payment_method: {
                        card: cardElement,
                    }
                }
            );

            if (error) {
                document.getElementById('error-message').textContent = error.message;
            } else {
                document.getElementById('error-message').textContent = 'Payment successful!';
                // Optionally, redirect to a success page
            }
        });
    </script>
</body>
</html>



Error Handling

    Retries: Implement retry logic for network-related errors.
    Logging: Log all requests and responses for monitoring and troubleshooting.

Security Considerations

    Ensure all API requests to Stripe are made over HTTPS.
    Handle sensitive information like API keys securely.
    Use Stripe's webhook secret to verify the integrity of webhook requests.

Testing

    Use Stripe's test mode for development and testing.
    Verify that payments are processed correctly and webhooks are handled appropriately.

By following these steps and using the code snippets provided, you can integrate Stripe into your Flask application to process payments for your lottery/betting system. Adjust configurations and endpoints as needed for your specific use case.

File: noble-shipping-nextjs/src/lib/stripe.ts
/**
 * Stripe utility module for payment processing
 * Handles payment intents, refunds, and webhook events
 */

import Stripe from 'stripe'

// Initialize Stripe client
if (!process.env.STRIPE_SECRET_KEY) {
    throw new Error('STRIPE_SECRET_KEY environment variable is required')
}

const stripe = new Stripe(process.env.STRIPE_SECRET_KEY, {
    apiVersion: '2024-06-20',
})

/**
 * Create a payment intent for processing payments
 */
export async function createPaymentIntent(
    amount: number,
    currency: string,
    metadata: Record<string, string>
): Promise<Stripe.PaymentIntent> {
    try {
        const paymentIntent = await stripe.paymentIntents.create({
            amount,
            currency,
            metadata,
            automatic_payment_methods: {
                enabled: true,
            },
        })
        return paymentIntent
    } catch (error) {
        console.error('Payment intent creation error:', error)
        throw new Error('Failed to create payment intent')
    }
}

/**
 * Retrieve payment intent details
 */
export async function retrievePaymentIntent(
    paymentIntentId: string
): Promise<Stripe.PaymentIntent> {
    try {
        const paymentIntent = await stripe.paymentIntents.retrieve(paymentIntentId)
        return paymentIntent
    } catch (error) {
        console.error('Payment intent retrieval error:', error)
        throw new Error('Failed to retrieve payment intent')
    }
}

/**
 * Process a refund for a payment
 */
export async function refundPayment(
    paymentIntentId: string,
    amount?: number
): Promise<Stripe.Refund> {
    try {
        const refund = await stripe.refunds.create({
            payment_intent: paymentIntentId,
            amount,
        })
        return refund
    } catch (error) {
        console.error('Refund processing error:', error)
        throw new Error('Failed to process refund')
    }
}

/**
 * Construct and verify webhook event
 */
export function constructWebhookEvent(
    payload: string | Buffer,
    signature: string
): Stripe.Event {
    const webhookSecret = process.env.STRIPE_WEBHOOK_SECRET
    if (!webhookSecret) {
        throw new Error('Stripe webhook secret not configured')
    }

    try {
        const event = stripe.webhooks.constructEvent(
            payload,
            signature,
            webhookSecret
        )
        return event
    } catch (error) {
        console.error('Webhook verification error:', error)
        throw new Error('Failed to verify webhook signature')
    }
}

/**
 * Export Stripe client for advanced usage
 */
export { stripe }
